generate-geodatabase-replica-from-feature-service/src/main/java/com/esri/arcgismaps/sample/generategeodatabasereplicafromfeatureservice/MainActivity.kt

```diff
@@ -0,0 +1,256 @@
+                return@launch
+            }
+            // create a geodatabase sync task with the feature service url
+            // This feature service shows a web map of portland street trees,
+            // their attributes, as well as related inspection information
+            val geodatabaseSyncTask = GeodatabaseSyncTask(getString(R.string.feature_server_url))
+            geodatabaseSyncTask.load().onFailure {
+                // if the metadata load fails, show the error and return
+                showError("Failed to fetch geodatabase metadata")
+                return@launch
+            }
+            // enable the generate button and add an onClickListener
+            generateButton.isEnabled = true
+            generateButton.setOnClickListener {
+                mapView.visibleArea?.let { polygon ->
+                    // start the geodatabase generation process
+                    generateGeodatabase(geodatabaseSyncTask, map, polygon.extent)
+                }
+            }
+        }
+    }
+
+    /**
+     * Starts a GeodatabaseSyncTask, runs a GenerateGeodatabaseJob and saves
+     * the geodatabase file into local storage
+     *
+     * @param geodatabaseSyncTask the GeodatabaseSyncTask object
+     * @param map mapView to render the feature layers on
+     * @param extents the selected extents of the mapView
+     */
+    private fun generateGeodatabase(geodatabaseSyncTask: GeodatabaseSyncTask, map: ArcGISMap, extents: Envelope) {
+        // clear any layers already on the map
+        map.operationalLayers.clear()
+        // clear all symbols drawn
+        graphicsOverlay.graphics.clear()
+
+        // create a boundary representing the extents selected
+        val boundary = Graphic(extents, boundarySymbol)
+        // add this boundary to the graphics overlay
+        graphicsOverlay.graphics.add(boundary)
+
+        lifecycleScope.launch {
+            // create generate geodatabase parameters for the selected extents
+            val defaultParameters = geodatabaseSyncTask.createDefaultGenerateGeodatabaseParameters(extents).getOrElse {
+                // show the error and return if the task fails
+                showError("Error creating geodatabase parameters")
+                return@launch
+            }
+
+            // set return attachments option to false
+            defaultParameters.returnAttachments = false
+            // create a generate geodatabase job
+            geodatabaseSyncTask.createGenerateGeodatabaseJob(defaultParameters, geodatabaseFilePath).run {
+                // create a dialog to show the jobs progress
+                val dialog = createProgressDialog(this).apply {
+                    // show the dialog
+                    show()
+                    // allow it to be cancellable
+                    setCancelable(true)
+                    // disable dismissal when tapped outside the dialog
+                    setCanceledOnTouchOutside(false)
```

Comment:
`setCanceledOnTouchOutside()` is a member of the `AlerDialog` class. I've also added a `setOnCancelListener` to the builder, since upon testing I noticed that cancelling the dialog by the back button doesn't trigger the `setNegativeButton` listener.

Reply:
When removing `setCanceledOnTouchOutside`, the cancel listener does get called when pressing the back button for me and I get the expected message: 
> Error fetching geodatabase: User canceled.

Another option would be to `setCancelable(true)`, so the job can only be canceled on cancel button click and not on tapping outside the dialog or a back button press.

Reply:
The `onCancelListener` does get called as I added that in addition to a `negativeButtonListener`. I think instead setting the dialog to `setCancelable(false)` works better.